### PR TITLE
Track Request ID, Response ID, and Response Code + Add response-code-specific error messages

### PR DESCRIPTION
**Pull request by deeptivaidymsft**
For [Error Improvements in VM Extensions](https://microsoft.sharepoint.com/:w:/t/ComputeVM/ERQYqh3QTktFrdtAgBMZvAgBLKT8DqKH3K0rKznDVCr11A?e=bJE1Zm) , when downloading scripts we will log the request ID, response ID, and response code (if provided) and give specific error messages for status codes 401, 404, 400, 500.

### CODE REVIEW

File: pkg/download/blob_test.go [hunk 1]
```diff
@@ -145,14 +151,65 @@ func Test_blobDownload_actualBlob(t *testing.T) {
...
+	_, body, err := Download(testctx, d)
 	require.Nil(t, err)
 	defer body.Close()
 	b, err := ioutil.ReadAll(body)
 	require.Nil(t, err)
 	require.EqualValues(t, chunk, b, "retrieved body is different body=%d chunk=%d", len(b), len(chunk))
 }
 
+func Test_blobDownload_fails_actualBlob404(t *testing.T) {
+	acct := os.Getenv("AZURE_STORAGE_ACCOUNT")
```
Comment (jscalev) [inline pkg/download/blob_test.go:163]: Are these set when our tests run in the lab?

File: pkg/download/blob_test.go [hunk 2]
```diff
@@ -145,14 +151,65 @@ func Test_blobDownload_actualBlob(t *testing.T) {
...
+	code, _, err := Download(testctx, d)
+	require.NotNil(t, err)
+	require.Equal(t, code, http.StatusNotFound)
+	require.Contains(t, err.Error(), "because it does not exist")
+	require.Contains(t, err.Error(), "Not Found")
+	require.Contains(t, err.Error(), "Service request ID")
+}
+
+func Test_blobDownload_fails_actualBlob409(t *testing.T) {
+	// before running this test, go to your storage account on portal > Configuration and disable Blob public access
```
Comment (jscalev) [inline pkg/download/blob_test.go:188]: It's great that you tested with actual blobs, but the unit tests will need to mock these failures or they'll fail in the lab and on others' machines

File: pkg/download/blobwithmsitoken.go [hunk 3]
```diff
@@ -55,6 +56,7 @@ func (self *blobWithMsiToken) GetRequest() (*http.Request, error) {
 		return nil, err
 	}
 
+	request.Header.Set(xMsClientRequestIdHeaderName, uuid.New().String())
```
Comment (jscalev) [inline pkg/download/blobwithmsitoken.go:59]: should trace the requestId

File: pkg/download/blobwithmsitoken_test.go [hunk 4]
```diff
@@ -54,6 +55,23 @@ func Test_realDownloadBlobWithMsiToken(t *testing.T) {
 	require.Contains(t, string(bytes), stringToLookFor)
 }
 
+func Test_realDownloadBlobWithMsiToken404(t *testing.T) {
+	if msiJson == "" || blobUri == "" || stringToLookFor == "" {
```
Comment (jscalev) [inline pkg/download/blobwithmsitoken_test.go:59]: see comments above

File: pkg/download/downloader.go [hunk 5]
```diff
@@ -61,20 +67,53 @@ func Download(downloader Downloader) (int, io.ReadCloser, error) {
...
+	errString := ""
+	requestId := response.Header.Get(xMsServiceRequestIdHeaderName)
 	switch downloader.(type) {
 	case *blobWithMsiToken:
 		switch response.StatusCode {
 		case http.StatusNotFound:
-			notFoundError := fmt.Errorf("Make sure Azure blob '%s' and managed identity exist, and identity has been given access to storage blob's container with 'Storage Blob Data Reader' role assignment. In case of user assigned identity, make sure you add it under VM's identity. For more info, refer https://aka.ms/RunCommandManagedLinux", request.URL.Opaque)
-			return response.StatusCode, nil, errors.Wrapf(notFoundError, MsiDownload404ErrorString)
-		case http.StatusForbidden:
+			notFoundError := fmt.Sprintf("RunCommand failed to download blob '%s' and recieved response code '%s'. Make sure Azure blob and managed identity exist, and identity has access to storage blob's container with 'Storage Blob Data Reader' role assignment. For user assigned identity, add it under VM's identity. For more info, see https://aka.ms/RunCommandManagedLinux", request.URL.Opaque, response.Status)
```
Comment (jscalev) [inline pkg/download/downloader.go:76]: nit: received

Comment (jscalev) [reply pkg/download/downloader.go:76]: There are a lot of grammar mistakes here and in the other errors. I recommend running them through ChatGPT (Please correct my grammar. {text}) to correct them.

File: pkg/download/url.go [hunk 6]
```diff
@@ -17,7 +24,11 @@ func NewURLDownload(url string) Downloader {
 
 // GetRequest returns a new request to download the URL
 func (u urlDownload) GetRequest() (*http.Request, error) {
-	return http.NewRequest("GET", u.url, nil)
+	req, err := http.NewRequest("GET", u.url, nil)
+	if req != nil {
+		req.Header.Add(xMsClientRequestIdHeaderName, uuid.New().String())
```
Comment (jscalev) [inline pkg/download/url.go:29]: trace

**Review by jscalev — CHANGES_REQUESTED** [review]

**Review by malhotrarohit — APPROVED** [review]

**Review by jscalev — APPROVED** [review]

**Pull request merged**